### ISSUE TITLE
Fix bluetooth

### ISSUE THREAD

**Issue by TheMeaningfulEngineer** [getsenic/senic-hub#394]
[Hub branch](https://github.com/getsenic/senic-hub/tree/nrf52-onboarding-nuimo)
[Os branch](https://github.com/getsenic/senic-os/tree/porting-the-app-stack)

## Status

[Senic hub commit](https://github.com/getsenic/senic-hub/commit/cbab10241e11ddef9d4b83a01815bd45107e514f)
~~[Os commit](https://github.com/getsenic/senic-os/commit/8e974ca0e4a8ab8a73b40d994b9ab8de83fe1159)~~ OS is set to always take the latest commit of the senic-hub branch. For reproducibility you must manually set the commit in the recipe.

Nots: 
* Showing networks in the app might take some time (limit 20sec)
* Nuimo and peripherals detection can take some time (limit 20sec)

#### Solved by https://github.com/getsenic/senic-hub/pull/411 [** awaiting review **]

~~Random~~ success

- OK Connecting to wifi
- ~~Sometimes~~ OK Detect and connect first nuimo
- OK Detect Sonos and Hue and autenticate
- OK Reach screen after the video tutorial and control the compoenents with nuimo

When fails, the device can be discovered but not connected to:

	nuimoctl --discover --adapter hci0
	nuimoctl --adapter hci0 --connect e0:88:72:c4:49:c2

resterting bluetooth reanables discovery (and a nuimo in the app is shown):

	systemctl bluetooth restart
	nuimoctl --discover --adapter hci0
	nuimoctl --adapter hci0 --connect e0:88:72:c4:49:c2

#### Dev tip

For components that don't require authentication, this is a 
shortcut to the vanilla onboarding without the mobile app

	# Hub
	nmcli dev wifi con SSID  password PASSWORD  ifname wlan0

	# PC
	curl -X POST   http://10.10.10.159:6543/-/nuimos
	curl -X POST   http://10.10.10.159:6543/-/config
	curl -X GET   http://10.10.10.159:6543/-/confnuimos

### Nuimo reconnect https://github.com/getsenic/senic-hub/issues/407

Random success

Precondition: vanilla onboarding

- After a onboarded first nuimo, go out of range
and reconnect when back in range.

The nuimo reconnects sometimes but at certain point
stops reconnecting.

### Power board on off https://github.com/getsenic/senic-hub/issues/408

Precondition: vanilla onboarding

- Once the board is powered up the nuimo is 
reconnected and can control the devices

**Comment by TheMeaningfulEngineer** [comment]
There is a lot going on with this issue. It could be multiple issues masked as one and at the moment  difficult to isolate.

It might be related to [this action](https://github.com/getsenic/senic-hub/blob/master/senic_hub/nuimo_app/__main__.py#L84) and caused just by the new module not connecting properly after it.
**TODO** Need to confirm the stack working with the dongle to confirm this assumption.

I've started a branch https://github.com/getsenic/senic-hub/tree/improve-logging
to in parallel diagnose this issue and improve the logging inside the app stack.

Given that multiple small changes might be introduced to the os for fixing this, here is the tracking os branch: https://github.com/getsenic/senic-os/tree/porting-the-app-stack

**Comment by TheMeaningfulEngineer** [comment]
This is the latest reproducible manifestation of this issue:
```
Exception in thread Thread-3:                                                                                                                                                                               
Traceback (most recent call last):                                                                                                                                                                          
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner                                                                                                                                     
    self.run()                                                                                                                                                                                              
  File "/usr/lib/python3.5/threading.py", line 862, in run                                                                                                                                                  
    self._target(*self._args, **self._kwargs)                                                                                                                                                               
  File "/home/root/senic-hub/senic_hub/nuimo_app/__init__.py", line 302, in check_nuimo_connection                                                                                                          
    if self.controller and self.controller.is_connected() is False and self.connection_failed is True:                                                                                                      
  File "/usr/lib/python3.5/site-packages/gatt/gatt_linux.py", line 363, in is_connected                                                                                                                     
    return self._properties.Get('org.bluez.Device1', 'Connected') == 1                                                                                                                                      
  File "/usr/lib/python3.5/site-packages/dbus/proxies.py", line 145, in __call__                                                                                                                            
    **keywords)                                                                                                                                                                                             
  File "/usr/lib/python3.5/site-packages/dbus/connection.py", line 651, in call_blocking                                                                                                                    
    message, timeout)                                                                                                                                                                                       
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.UnknownObject: Method "Get" with signature "ss" on interface "org.freedesktop.DBus.Properties" doesn't exist                                      
                                                                                                                                                                                                            
                                                                                                                                                                                                            
2018-01-30 10:54:04,687 CRITI [senic_hub.nuimo_app] Connection failed e0:88:72:c4:49:c2: Device does not exist, check adapter name and MAC address.                                                         
2018-01-30 10:54:04,722 INFO  [senic_hub.nuimo_app] Connecting to Nuimo controller e0:88:72:c4:49:c2...                                                                                                     
2018-01-30 10:54:04,748 CRITI [senic_hub.nuimo_app] Connection failed e0:88:72:c4:49:c2: Device does not exist, check adapter name and MAC address.                                                         
2018-01-30 10:54:04,766 INFO  [senic_hub.nuimo_app] Connecting to Nuimo controller e0:88:72:c4:49:c2...                                                                                                     
2018-01-30 10:54:04,777 CRITI [senic_hub.nuimo_app] Connection failed e0:88:72:c4:49:c2: Device does not exist, check adapter name and MAC address.                                                         
2018-01-30 10:54:04,794 INFO  [senic_hub.nuimo_app] Connecting to Nuimo controller e0:88:72:c4:49:c2...                                                                                                     
2018-01-30 10:54:04,805 CRITI [senic_hub.nuimo_app] Connection failed e0:88:72:c4:49:c2: Device does not exist, check adapter name and MAC address.                                                         
2018-01-30 10:54:04,822 INFO  [senic_hub.nuimo_app] Connecting to Nuimo controller e0:88:72:c4:49:c2...                                                                                                     
2018-01-30 10:54:04,847 CRITI [senic_hub.nuimo_app] Connection failed e0:88:72:c4:49:c2: Device does not exist, check adapter name and MAC address.                                                         
2018-01-30 10:54:04,865 INFO  [senic_hub.nuimo_app] Connecting to Nuimo controller e0:88:72:c4:49:c2...                                                                                                     
2018-01-30 10:54:04,876 CRITI [senic_hub.nuimo_app] Connection failed e0:88:72:c4:49:c2: Device does not exist, check adapter name and MAC address.                                                         
2018-01-30 10:54:04,894 INFO  [senic_hub.nuimo_app] Connecting to Nuimo controller e0:88:72:c4:49:c2...                                                                                                     
2018-01-30 10:54:04,904 CRITI [senic_hub.nuimo_app] Connection failed e0:88:72:c4:49:c2: Device does not exist, check adapter name and MAC address.                                                         
2018-01-30 10:54:04,922 INFO  [senic_hub.nuimo_app] Connecting to Nuimo controller e0:88:72:c4:49:c2...                                                                                                     
2018-01-30 10:54:04,932 CRITI [senic_hub.nuimo_app] Connection failed e0:88:72:c4:49:c2: Device does not exist, check adapter name and MAC address.                                                         
2018-01-30 10:54:04,950 INFO  [senic_hub.nuimo_app] Connecting to Nuimo controller e0:88:72:c4:49:c2...                                                                                                     
2018-01-30 10:54:04,960 CRITI [senic_hub.nuimo_app] Connection failed e0:88:72:c4:49:c2: Device does not exist, check adapter name and MAC address.                                                         
2018-01-30 10:54:04,977 INFO  [senic_hub.nuimo_app] Connecting to Nuimo controller e0:88:72:c4:49:c2...                                                                                                     
2018-01-30 10:54:04,988 CRITI [senic_hub.nuimo_app] Connection failed e0:88:72:c4:49:c2: Device does not exist, check adapter name and MAC address.                                                         
2018-01-30 10:54:05,005 INFO  [senic_hub.nuimo_app] Connecting to Nuimo controller e0:88:72:c4:49:c2...                                                                                                     
2018-01-30 10:54:05,015 CRITI [senic_hub.nuimo_app] Connection failed e0:88:72:c4:49:c2: Device does not exist, check adapter name and MAC address.                                                         
Fatal Python error: Cannot recover from stack overflow.                                                                                                                                                     
```  

It is being addressed in this branch:
https://github.com/getsenic/senic-hub/tree/nrf52-onboarding-nuimo

The repo with the os related changes has been rebased to the os master,
and will be PRed once the issue is finished.
https://github.com/getsenic/senic-os/commits/porting-the-app-stack

**Comment by TheMeaningfulEngineer** [comment]
Vanilla onboarding and Power board on off solved with this.
https://github.com/getsenic/senic-hub/pull/421

Reconnects will be addressed in a separate issue.
